### DISCUSSION TITLE
Pasteboard sandboxing on Apple Silicon?

### DISCUSSION THREAD

**Discussion by DominusExult** [jralls/gtk-osx-build#92] [Q&A] [answered]
Hi all,
We have a GTK3 based application from which we drag onto an SDL based application (Exult Studio and Exult from https://exult.info). This works by getting the public.file-url from the pasteboard and works fine on Intel macs.
On Apple Silicon the public.file-url is empty when Exult receives the drop from Exult Studio (the log says "Couldn't get a copy of a URL-flavored data from the pasteboard." "Looked for URLs on the pasteboard, but found none."). Pasteboard Viewer app also shows an empty entry for public.file-url and SDL's test app shows "Failed to get a sandbox extension" when we drop from Exult Studio.
If I use an Intel compiled Exult Studio everything works again (even with an Arm64 compiled Exult).

From https://wadetregaskis.com/mac-app-sandboxing-interferes-with-drag-drop/ (which I found when trying to get behind this bug) it seemed to me that only "receiving" apps could run into this sandboxing problem/bug not the emitter (and on top of that neither are sandboxed per se, but it seems Apple Silicon macOS has tightened security). But in our case it seems Exult Studio (emitter) is causing the problem.

Why post this here? Maybe someone has an idea *OR* can anyone recommend a GTK3 application that runs on Apple Silicon *and* has some sort of drag and drop. Just so I can try and see if it is an Exult only problem or maybe a bigger GTK3 issue.

**Comment by jralls** [comment, marked as answer]
It might depend on how you're making the file URL to put on the pasteboard. https://gitlab.gnome.org/GNOME/gtk/-/issues/6480, fixed last week, was about copying from the generic file chooser but the underlying cause, a mismatch in the NSPasteboardType, might have some bearing on your problem. I suggest that you build Gtk3 from git (you can use the `gtk-3.0` module in either `modulesets` or `modulesets-unstable`) and test. In my testing of the fix I successfully dragged a file entry from a non-native GtkFileChooser window to TextEdit, which opened the dragged file as one would expect.

**Reply by DominusExult** [reply]
Oh wow, so very recent!
I'll try my hands tomorrow, if this was the fix, I'll cheer!

**Reply by DominusExult** [reply]
This issue is indeed fixed by Gtk3 git. Thank you so much! (I guess this gets ported to Gtk4 if needed there as well)

**Reply by jralls** [reply]
Good, thanks for following up.

It doesn't apply at all to Gtk4. The macOS backend in Gtk4 is a complete rewrite from scratch.

**Reply by DominusExult** [reply]
I *was* wondering why my Intel build still worked on both Intel and Apple Silicon machines but it turns out that was because I was targeting macOS 10.11 with that :)
